File: actions.py
STEP = 1
X = 0
Y = 1


def _move(position, step):
  x = position[X]
  y = position[Y] + step
  return (x, y)


class Backward:
  def move(self, position):
    return _move(position, -STEP)

  def turn(self, direction):
    return direction


class Forward:
  def move(self, position):
    return _move(position, +STEP)

  def turn(self, direction):
    return direction


class Left:
  def move(self, position):
    return position

  def turn(self, direction):
    if direction is 'N':
      return 'W'
    elif direction is 'W':
      return 'S'
    elif direction is 'S':
      return 'E'
    elif direction is 'E':
      return 'N'


class Right:
  def move(self, position):
    return position

  def turn(self, direction):
    if direction is 'N':
      return 'E'
    elif direction is 'E':
      return 'S'
    elif direction is 'S':
      return 'W'
    elif direction is 'W':
      return 'N'

File: actions.ts
import { Direction } from './directions';

export type Position = [number, number];

const STEP = 1;
const Y = 1;

function _move(position: Position, step: number): Position {
  const newPosition = position.slice() as Position;
  newPosition[Y] += step
  return newPosition;
}

function step(mode: string): number {
  return mode == 'turbo' ? 2 * STEP : STEP;
}

export interface Action {
  move(position: Position, mode: string): Position;
  turn(direction: Direction): Direction;
}

export class Backward implements Action {
  public move(position: Position, mode: string): Position {
    return _move(position, -step(mode));
  }

  public turn(direction: Direction): Direction {
    return direction;
  }
}

export class Forward implements Action {
  public move(position: Position, mode: string): Position {
    return _move(position, step(mode));
  }

  public turn(direction: Direction): Direction {
    return direction;
  }
}

export class Left implements Action {
  public move(position: Position, mode: string): Position {
    return position;
  }

  public turn(direction: Direction): Direction {
    return direction.left();
  }
}

export class Right implements Action {
  public move(position: Position, mode: string): Position {
    return position;
  }

  public turn(direction: Direction): Direction {
    return direction.right();
  }
}
